mldsa/reduce.h

```diff
@@ -10,12 +10,27 @@
 #include "params.h"
 
 #define MONT -4186625 /* 2^32 % MLDSA_Q */
-#define QINV 58728449 /* q^(-1) mod 2^32 */
 #define REDUCE_DOMAIN_MAX (INT32_MAX - (1 << 22))
 #define REDUCE_RANGE_MAX 6283009
+#define MONTGOMERY_REDUCE_DOMAIN_MAX (MLDSA_Q * (1LL << 31))
 
 #define montgomery_reduce MLD_NAMESPACE(montgomery_reduce)
-int32_t montgomery_reduce(int64_t a);
+/*************************************************
+ * Name:        montgomery_reduce
+ *
+ * Description: For finite field element a with
+ *              -2^{31}MLDSA_Q <= a <= MLDSA_Q*2^31,
+ *              compute r \equiv a*2^{-32} (mod MLDSA_Q) such that
+ *              -MLDSA_Q < r < MLDSA_Q.
+ *
+ * Arguments:   - int64_t: finite field element a
+ *
+ * Returns r.
+ **************************************************/
+int32_t montgomery_reduce(int64_t a)
+__contract__(
+  requires(a >= -MONTGOMERY_REDUCE_DOMAIN_MAX && a <= MONTGOMERY_REDUCE_DOMAIN_MAX)
+);
```

Comment:
Declare _and_ define here for inlining into call sites?

Reply:
Let's do that later when benchmarking is in place.